File: benchmarks/executable/restaurant_f64_strict.c
// C23 reference for the Restaurant strict binary64 arithmetic workload.

#include <stdio.h>

#ifdef _WIN32
#include <fcntl.h>
#include <io.h>
#endif

/*
 * Volatile inputs keep the floating-point operations and NaN comparison in
 * the executable path. The release recipe preserves strict IEEE semantics.
 */
static volatile double sum_left = 1.5;
static volatile double sum_right = 2.25;
static volatile double difference_left = 9.5;
static volatile double difference_right = 5.5;
static volatile double product_left = 1.5;
static volatile double product_right = 2.0;
static volatile double quotient_left = 7.5;
static volatile double quotient_right = 2.5;
static volatile double signed_zero_input = 0.0;
static volatile double nan_left = 0.0;
static volatile double nan_right = 0.0;

static int strict_float_check(void) {
    const double sum = sum_left + sum_right;
    const double difference = difference_left - difference_right;
    const double product = product_left * product_right;
    const double quotient = quotient_left / quotient_right;
    const double signed_zero = -signed_zero_input;
    const double nan = nan_left / nan_right;

    return sum == 3.75 &&
           difference == 4.0 &&
           product == 3.0 &&
           quotient != 4.0 &&
           sum > 3.0 &&
           sum >= 3.75 &&
           quotient < 4.0 &&
           quotient <= 3.0 &&
           signed_zero == 0.0 &&
           nan != nan;
}

int main(void) {
#ifdef _WIN32
    if (_setmode(_fileno(stdout), _O_BINARY) == -1) return 1;
#endif

    if (strict_float_check()) {
        if (fputs("Float strict ok\n", stdout) == EOF) return 1;
    } else if (fputs("Float strict bad\n", stdout) == EOF) {
        return 1;
    }
    return 0;
}

File: benchmarks/executable/restaurant_f64_strict.rs
// Rust 2024 reference for the Restaurant strict binary64 arithmetic workload.

use std::hint::black_box;

fn strict_float_check() -> bool {
    let sum = black_box(1.5_f64) + black_box(2.25_f64);
    let difference = black_box(9.5_f64) - black_box(5.5_f64);
    let product = black_box(1.5_f64) * black_box(2.0_f64);
    let quotient = black_box(7.5_f64) / black_box(2.5_f64);
    let signed_zero = -black_box(0.0_f64);
    let nan = black_box(0.0_f64) / black_box(0.0_f64);

    sum == 3.75 &&
        difference == 4.0 &&
        product == 3.0 &&
        quotient != 4.0 &&
        sum > 3.0 &&
        sum >= 3.75 &&
        quotient < 4.0 &&
        quotient <= 3.0 &&
        signed_zero == 0.0 &&
        nan != nan
}

fn main() {
    if strict_float_check() {
        println!("Float strict ok");
    } else {
        println!("Float strict bad");
    }
}
